lib/lab_manager/provider/vsphere.rb

```diff
@@ -136,6 +142,65 @@ def power_on
       end
     end
 
+    def shutdown_vm(opts = {})
+      fail ArgumentError, 'Virtual machine data not present' unless instance_uuid
+
+      VSphere.connect.with do |vs|
+        Retryable.retryable(tries: 6) do
+          server = vs.servers.get(instance_uuid)
+          fail VmNotExistsError, 'Vm not exists!' unless server
+          break if server.power_state == 'poweredOff'
+
+          case opts[:mode] || 'managed'
+          when 'hard'
+            server.stop(force: true)
+          when 'soft'
+            server.stop(force: false)
+          when 'managed'
+            begin
+              ts = server.stop(force: false)
+            rescue => e
+              LabManager.logger.warn "gracefully shut down the machine failed, "\
+                "trying force off, #{e.to_s}"
+              ts = server.stop(force: true)
+            end
+          end
+
+          Retryable.retryable(tries: 10, sleep: 1) do
+            fail ShutdownVmError, '' unless  
+              vs.get_virtual_machine(server.id)['power_state'] == 'poweredOff'
+          end
```

Comment:
how does work `force:false`? Does it only send signal to vmtool and machine starts to shut down? In that case 10s shudn't be enough.